### PR TITLE
fix(google-content-sync): harden against the poison-page class, make the snapshot obligation durable, type-check the lambda

### PR DESCRIPTION
**Pull request by krishagel**
## Summary

Hardens the Google Drive content-sync pipeline against the whole *class* of poison-page failures that caused the 2026-08-06 prod incident, closes the durable-obligation gap around selection snapshots, and puts the lambda under the TypeScript compiler for the first time.

> **Stacks on #1617.** This branch is cut from `claude/google-sync-queue-backlog-b836ae`; until #1617 merges, its two commits appear in this diff. Merge #1617 first, then this.

Follow-up scope approved by @hagelk after #1617 deliberately shipped root-cause-only.

## Gap 1 — all-or-nothing page parsing (`lib/repositories/google-drive/drive-client.ts`)

`changesListSchema`/`filesListSchema` used to `z.array(entrySchema).parse()` the whole payload: one malformed entry among 1000 rejected the page, the cursor never advanced, and the connector stalled — the original `fileId` incident, generalized.

- **Per-entry validation**: list envelopes now parse entries as `unknown[]`, then validate one at a time (`parseDriveEntries`). A malformed member costs that one entry, not the page. Skipped entries surface as `skippedEntries` (index, best-effort id, zod issues) and are logged loudly at every call site. **A parse failure is never reinterpreted as a removal** — dropped entries suppress the unseen-source sweep instead (see Gap 2).
- **Validators loosened to Google's actual contract**, each with a comment tying it to its consumer: `modifiedTime`/`time` plain strings (consumers do `new Date(...)`); `size` plain string (the streaming byte bound is authoritative); `shortcutDetails.targetId`/`targetMimeType` optional (`resolveShortcut` already has the graceful "no target" path; `targetMimeType` has no reader); `.nullish()` + normalizing transforms where Google could send explicit `null`.
- **`webViewLink`/`iconLink`**: strict `.url()` dropped, but a non-http(s) scheme (`javascript:`, `data:`) is discarded rather than persisted, so the loosening cannot open an href-injection path.
- **`getFile()` (single-entity sibling of the same gap)**: now `safeParse`s and throws a typed `GoogleDriveUnreadableFileError` (fileId + issues) instead of a raw `ZodError`. Callers classify it explicitly: a change-processing failure fails that one record (`recordSourceFailure`, never `markSourceMissing`, cursor advances); an unreadable shortcut target or selection root during enumeration counts as a skipped entry. Found by adversarial review of the first commit — a ZodError escaping `getFile` mid-change-page would have replayed the page until the DLQ, the exact pattern this PR eliminates.

## Gap 2 — snapshot obligation now durable (`infra/lambdas/google-content-sync/`)

Previously, a change page that demanded a selection snapshot suppressed every per-page cursor write until the snapshot finished. A 900s Lambda timeout or budget failure threw away all change-page work and replayed it from the original cursor — indefinitely, if the snapshot kept failing.

- New `reconcile.ts` extracts the changes-feed loop against injected collaborators (same pattern as `changes.ts`/`safety.ts`; `index.ts` cannot be imported from the root jest gate). Ordering invariants, each behaviorally tested:
  1. The obligation is written durably (`metadata.selectionSnapshotPendingAt`, jsonb merge — **no migration needed**) *before* the cursor moves past the page that raised it.
  2. The cursor persists after every page, unconditionally.
  3. The flag clears only after a *complete* snapshot; an incomplete one (dropped entries) retains it, so the next run retries.
- On sync start, a pending obligation is discharged *before* consuming new pages (`resolveSyncCursor`).
- `shouldRetireUnseenSources`: `markUnseenSourcesMissing` only runs after a complete enumeration — a dropped entry says nothing about its file's existence, so retiring on partial data would be silent data loss.
- `reconcileInitial` records the obligation on an incomplete initial/410-rebuild too (`setSelectionSnapshotPending(context, !complete)`), aligning it with the changes-loop path (correctness-review finding).

## Gap 3 — lambda now type-checked

`infra/lambdas/google-content-sync/` had no tsconfig and was excluded from both the root and infra `tsc` runs — esbuild only transpiles. Added `tsconfig.json` (modeled on `unified-content-processor`) and `typecheck:google-content-sync`, wired into `infra`'s `build` script exactly like the sibling check. The duplicate local re-derivation of `GoogleDriveChange` in `index.ts` is replaced by the exported type.

## Deliberately unchanged

- No DLQ redrive, no alarm-threshold changes (locked decisions from #1617).
- `enumerateInitialFiles`'s 403/404-on-selection-root behavior (retire via the missing path) is pre-existing and intentional parity with file-level 403/404 semantics — not widened here.

## Verification

- `bun run lint` — zero warnings
- `bun run typecheck` + `typecheck:google-content-sync` — clean
- Full jest suite: 574 suites / 5465 passed (one unrelated `document-generation-service` xlsx test hit its 5s timeout under machine load; passes in isolation in 1.1s)
- Full local Playwright gate: 306 passed / 40 skipped; one `nexus/model-router` failure + 10 flaky-passed-on-retry, all Nexus/Atrium UI specs untouched by this backend diff. The failed spec was rerun in isolation and passed (10 passed / 1 flaky-passed-on-retry / 0 failed) — load-induced flake from two sessions' gates sharing the machine, not a regression
- New coverage: 16 drive-client contract tests (per-entry skip, null tolerance, scheme filtering, shortcut edge cases, typed `getFile` error), 12 reconcile-loop behavior tests (cursor/obligation ordering, incomplete-snapshot retention), change-scope wiring assertions

**Named E2E flows**: none — backend Lambda path, not reachable via Playwright; covered by unit/contract tests above.

### CODE REVIEW

File: infra/lambdas/google-content-sync/index.ts [hunk 1]
```diff
@@ -1328,16 +1445,73 @@ async function retryDeferredDownloads(
...
+        ? "Shared Drive removed from the change feed"
+        : "Skipped non-file Google Drive change entry",
+      {
+        connectorId: context.connector.id,
+        changeType: change.changeType ?? null,
+        driveId: change.driveId ?? null,
+        time: change.time ?? null,
+      },
+    );
+    return driveRemoved;
```
[the remark this answers is not in the document]
Comment (krishagel) [reply infra/lambdas/google-content-sync/index.ts:1546]: Confirmed and fixed in `47ccea8`.

Traced it: `processGoogleDriveChange` returns `true` on a drive removal to request a snapshot, but `enumerateInitialFiles` rethrew 403/404 when `selectionKind === "drive"`, so that snapshot never completed — `handleSyncFailure` → `markConnectorAccessLost`, which updates every `repositoryConnectorSources` row for the connector, and `applyDeletionGrace` then removes content belonging to file and folder selections that were still perfectly readable.

Took your first option (handle it within the snapshot) rather than retiring by `driveId`. The missing-error branch no longer excludes `drive` selections: the selection is counted in `inaccessibleSelectionCount`, logged with its `selectionKind`, and enumeration continues, leaving retirement to `markUnseenSourcesMissing`. That is what the change handler already documented as the recovery path, and because the sweep is scoped to what this enumeration *did not see*, it also handles the case retiring by `driveId` would get wrong — a file that lives in the removed drive but is individually reachable through its own file selection stays alive.

Two things I checked before widening the branch:

- **Total access loss still converges.** Every selection lands in the same branch, the seen set is empty, and `markUnseenSourcesMissing` retires all sources with `status: "missing"`. `applyDeletionGrace` selects on `["missing", "access_lost"]`, so the same grace period applies. The one thing lost versus the old path is the `access_lost` *status* as a distinct signal on a fully revoked connector — flagging that explicitly in case you'd rather keep the distinction with an all-selections-inaccessible guard; I left it out as a redesign beyond this comment's scope.
- **The unseen sweep is still gated.** `shouldRetireUnseenSources` is unchanged, so a dropped entry anywhere in the enumeration still suppresses retirement entirely.

Also corrected the comment in `processGoogleDriveChange` that claimed a hard access loss surfaces through the 403/404 path — that was describing the behavior this change removes.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/repositories/google-drive/drive-client.ts [hunk 2]
```diff
@@ -24,50 +24,122 @@ const FILE_FIELDS = [
...
-  owners: z.array(driveUserSchema).optional().default([]),
-  shortcutDetails: shortcutDetailsSchema.optional(),
+  parents: z
+    .array(z.string())
+    .nullish()
+    .transform((value) => value ?? []),
+  driveId: optionalString,
+  // Plain string, not `.datetime()`: the only consumers are `new Date(...)`
+  // and a revision string. Callers guard invalid dates themselves.
+  modifiedTime: optionalString,
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/repositories/google-drive/drive-client.ts]: Confirmed and fixed in `aa079f4`.

The failure mode is exactly as described, and the "not counted as skipped" half is the part that makes it durable: `sourceIdentityFields` does `new Date(file.modifiedTime)` into a `timestamp` column, the import transaction fails, and because the entry *validated* it never lands in `skippedEntries` — so `shouldRetireUnseenSources` isn't suppressed and the changes-feed cursor advances past it. The file is then only retried if it happens to change again.

Took the "normalize to `undefined`" option rather than restoring datetime validation, since rejecting the entry is the poison-page class this PR exists to remove. New `optionalTimestamp` validator accepts any string, keeps only what `Date.parse` accepts, and normalizes the rest to `undefined`, which puts the field on its existing absent path (`modifiedTime: null` in the row, `sourceRevision` falling through to the next component).

Scope note: applied to `googleDriveFileSchema.modifiedTime` only. `driveChangeSchema.time` is the other relaxed timestamp, but its sole consumer is a log field in `processGoogleDriveChange` — never `new Date(...)` — so it stays a plain string.

Covered by a new contract test asserting both halves: the value is dropped, *and* the entry is still returned rather than skipped.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
🤖 pr-fix routine round 1 complete.

- Review comments addressed: 2 of 2 actionable (Codex P1 + P2, both confirmed real)
- CI failures fixed: 0 (all 10 checks were green before this round)
- Commits pushed: 2

| Finding | Commit | Fix |
|---|---|---|
| P1 — drive removal escalates to connector-wide access loss | `47ccea8` | `enumerateInitialFiles`' missing-error branch no longer excludes `drive` selections; retirement is scoped to the unseen sweep |
| P2 — unparseable `modifiedTime` becomes an Invalid Date | `aa079f4` | New `optionalTimestamp` validator normalizes unparseable values to `undefined` |

Two things worth a human eye:

1. **The PR description is now slightly stale.** "Deliberately unchanged" says `enumerateInitialFiles`' 403/404-on-selection-root behavior is "not widened here" — P1's fix does change it, for `drive` selections specifically. I left your prose alone rather than rewriting it; worth a one-line edit before merge.
2. **`access_lost` as a distinct status on a fully revoked connector.** With P1's fix, a connector whose every selection 403s now retires its sources as `missing` rather than `access_lost`. Content still converges to deleted (`applyDeletionGrace` covers both statuses), but the connector-level signal changes. Restoring the distinction would need an all-selections-inaccessible guard, which felt like a redesign beyond the review comment — flagging rather than deciding it.

**Verification caveat — read before merging.** I could not run the local gate: `bun install` fails in this routine's sandbox with `403` on npm tarball fetches (`registry.npmjs.org` is blocked by the egress policy here), so `lint` / `typecheck` / `jest` did not run locally. The two new tests were validated by executing their source-slice assertions directly against `index.ts`, and both fixes are small and localized, but **CI is the authoritative gate for this round** — do not merge on my say-so alone if the checks below are not green.

The Claude review comment on this PR was read and needed no code changes: its three "minor / non-blocking" items (the `persistSnapshotCursor` operation-label change, the brittleness of source-inspection tests, and the redundant trailing `.optional()` on `optionalString`) are observations and style notes, not defects. The first is the only one with any operational edge — worth confirming no dashboard keys off that operation name.



---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
🤖 pr-fix routine round 2 complete.

- Review comments addressed: 1 of 1 actionable (the `resolveShortcut` finding from the round-1 re-review)
- CI failures fixed: 0 (all 11 checks were green on `47ccea8`)
- Commits pushed: 1

**The `resolveShortcut` finding is correct and I've fixed it in `cde8217`.** Verified both halves of the claim against the source:

- `add()` (`index.ts`) catches only `GoogleDriveUnreadableFileError` and rethrows everything else; the per-selection catch in `enumerateInitialFiles` classifies only that type and `isGoogleDriveMissingError`. A bare `Error` reaches `throw error` and takes down the whole selection snapshot.
- `recordDriveChangeFailure` falls through to `if (!isGoogleDriveMissingError(error)) throw error;`, and `reconcileChangePages` persists the cursor only *after* the full page loop — so the throw lands first and the cursor never advances.

And it is genuinely reachable rather than theoretical: this PR is what loosened `shortcutDetails.targetId` from `z.string()` to optional, so a shortcut whose target was deleted now parses cleanly and then throws deterministically on every subsequent run.

Fix is the suggested one — `resolveShortcut` throws `GoogleDriveUnreadableFileError` with a `shortcutDetails.targetId` issue, so both call sites route it through paths they already have: a per-entry skip during enumeration, `recordSourceFailure` with an advancing cursor during change processing. Also corrected the schema comment and the test comment that both asserted the "no target" path was already graceful; that claim being wrong is what let this through.

**One tradeoff I made rather than decided for you:** the enumeration path counts this as a skipped entry, which suppresses `markUnseenSourcesMissing` for that run. For a *permanently* dangling shortcut that means retirement stays suppressed on that connector until the shortcut is removed. I took it deliberately — it is the existing conservative disposition for anything the client cannot resolve, and it fails toward stale content rather than deletion — but if you'd rather a dangling shortcut be a benign skip that doesn't block the sweep, that is a one-line change to make it not increment `skippedEntryCount`.

Round-1 verification caveat still applies: `bun install` is blocked by egress policy in this sandbox (403 on npm tarballs), so nothing was run locally. Round 1's changes were confirmed green by CI on `47ccea8` — **Test, Lint and Type Check passed**, along with all 11 checks. This round's commit needs the same confirmation before merge.



---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
🤖 pr-fix routine round 3 — nothing further to change; marking `pr-fix-done`.

**All 7 checks green on `cde8217`**, including Test, Lint and Type Check. The round-2 re-review came back with no blocking findings and independently confirms both round-1 fixes — it traces `GoogleDriveUnreadableFileError` as now classified consistently at every call site *including* `resolveShortcut`, and confirms drive-removal retirement is scoped to `markUnseenSourcesMissing` rather than `markConnectorAccessLost`.

Resolved the two Codex threads (P1 and P2), both fixed and replied to.

### What I did not change, and why

The round-2 review left three "minor / worth a look" notes. None are change requests, but one deserves your explicit call rather than my silent judgement:

**Envelope fields are still strict-parsed.** `filesListSchema` / `changesListSchema` validate `nextPageToken` / `newStartPageToken` through `optionalString` *before* per-entry parsing runs, so a non-string value in either field still throws the whole page pre-cursor-advance — the poison-page mode, narrowed from "any of 1000 entries" to "one of two envelope fields." The reviewer's read is that this is probably acceptable residual risk since both are opaque server-generated tokens, and asks whether it's deliberate.

I left it alone. It looks deliberate to me — you wrote `rawEntriesSchema` specifically for the entries and left the envelope strict — and it is a real narrowing of blast radius rather than an oversight. But I can't confirm your intent, and "make the envelope tolerant too" is a ~5-line change if you'd rather close the class completely. Worth one line in "Deliberately unchanged" either way.

The other two notes need no action: the `webViewLink`/`iconLink` scheme check being a prefix test rather than a full URL parse is correct for its stated goal (block `javascript:`/`data:`), and the `recordSkipped`/`recordUnreadable` logging similarity is too small to abstract.

### Standing caveats for whoever merges

1. **Nothing was verified locally.** `bun install` is blocked by egress policy in this routine's sandbox (403 on npm tarballs), so lint/typecheck/jest never ran here. Every claim above rests on CI, which is green on the current head.
2. **The PR description is stale in one spot.** "Deliberately unchanged" still says `enumerateInitialFiles`' 403/404-on-selection-root behavior is "not widened here" — the P1 fix does widen it, for `drive` selections. One-line edit before merge.
3. **Open question from round 2**, restated so it isn't lost: a permanently dangling shortcut now suppresses `markUnseenSourcesMissing` on that connector until the shortcut is removed. Deliberate (fails toward stale content, not deletion), reversible in one line if you disagree.
4. **This PR still stacks on #1617** — merge that first.

Removing `pr-fix-done` re-queues this PR for the routine.



---
_Generated by [Claude Code](https://claude.ai/code)_
